src/mono/msbuild/AndroidAppBuilder/Templates/runtime-android.c

```diff
@@ -0,0 +1,197 @@
+{
+    va_list args;
+    char *formatted = NULL;
+    va_start (args, msg);
+    vasprintf (&formatted, msg, args);
+    va_end (args);
+    return formatted;
+}
+
+static MonoObject *
+fetch_exception_property (MonoObject *obj, const char *name, bool is_virtual)
+{
+    MonoMethod *get = NULL;
+    MonoMethod *get_virt = NULL;
+    MonoObject *exc = NULL;
+
+    get = mono_class_get_method_from_name (mono_get_exception_class (), name, 0);
+    if (get) {
+        if (is_virtual) {
+            get_virt = mono_object_get_virtual_method (obj, get);
+            if (get_virt)
+                get = get_virt;
+        }
+
+        return (MonoObject *) mono_runtime_invoke (get, obj, NULL, &exc);
+    } else {
+        printf ("Could not find the property System.Exception.%s", name);
+    }
+
+    return NULL;
+}
+
+static char *
+fetch_exception_property_string (MonoObject *obj, const char *name, bool is_virtual)
+{
+    MonoString *str = (MonoString *) fetch_exception_property (obj, name, is_virtual);
+    return str ? mono_string_to_utf8 (str) : NULL;
+}
+
+void
+unhandled_exception_handler (MonoObject *exc, void *user_data)
+{
+    MonoClass *type = mono_object_get_class (exc);
+    char *type_name = strdup_printf ("%s.%s", mono_class_get_namespace (type), mono_class_get_name (type));
+    char *trace = fetch_exception_property_string (exc, "get_StackTrace", true);
+    char *message = fetch_exception_property_string (exc, "get_Message", true);
+    
+    LOG_ERROR("UnhandledException: %s %s %s", type_name, message, trace);
+
+    free (trace);
+    free (message);
+    free (type_name);
+    exit (1);
+}
+
+void
+log_callback (const char *log_domain, const char *log_level, const char *message, mono_bool fatal, void *user_data)
+{
+    LOG_INFO ("(%s %s) %s", log_domain, log_level, message);
+    if (fatal) {
+        LOG_ERROR ("Exit code: %d.", 1);
+        exit (1);
+    }
+}
+
+int
+mono_mobile_runtime_init (void)
+{
+    // for now, only Invariant Mode is supported (FIXME: integrate ICU)
+    setenv ("DOTNET_SYSTEM_GLOBALIZATION_INVARIANT", "1", true);
+    // uncomment for debug output:
+    //
+    // setenv ("MONO_LOG_LEVEL", "debug", TRUE);
+    // setenv ("MONO_LOG_MASK", "all", TRUE);
+
+    bool wait_for_debugger = false;
+    chdir (bundle_path);
+
+    // TODO: set TRUSTED_PLATFORM_ASSEMBLIES, APP_PATHS and NATIVE_DLL_SEARCH_DIRECTORIES
+    monovm_initialize(0, NULL, NULL);
+
+    mono_debug_init (MONO_DEBUG_FORMAT_MONO);
+    mono_install_assembly_preload_hook (assembly_preload_hook, NULL);
+    mono_install_unhandled_exception_hook (unhandled_exception_handler, NULL);
+    mono_trace_set_log_handler (log_callback, NULL);
+    mono_set_signal_chaining (true);
+    mono_set_crash_chaining (true);
+
+    if (wait_for_debugger) {
+        char* options[] = { "--debugger-agent=transport=dt_socket,server=y,address=0.0.0.0:55555" };
+        mono_jit_parse_options (1, options);
+    }
+    mono_jit_init_version ("mono.app", "mobile");
```

Comment:
nit: we used `dotnet.ios` for the iOS template, let's use `dotnet.android` here to be consistent.